### PR TITLE
improve locking handling for background jobs

### PR DESCRIPTION
**Pull request by mikehelmick**
## Proposed Changes

* add job lock to JWKS
* return success for other jobs if the reason for lock fail was the lock is already held
* goal - reduce alert noise 

**Release Note**

```release-note
Improve locking for background jobs: export-import, jwks, key rotation
```

### CODE REVIEW

**Review by sethvargo — COMMENTED** [review]
This seems good. I'm worried about a case where we fail to acquire the lock for many, many tries though. Could we have some upper bound or quiescence to alert if the lock is never acquired after some time? 

Specifically, I'm worried about a bug or crash that fails to abort.

**Review by sethvargo — APPROVED** [review]
/hold

**Comment by mikehelmick** [comment]
added open census metrics to indicate job success

File: internal/jwks/handle_update_all.go [hunk 1]
```diff
@@ -15,24 +15,45 @@
...
+
 func (s *Server) handleUpdateAll() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
-
 		logger := logging.FromContext(ctx).Named("handleUpdateAll")
 
+		unlock, err := s.manager.db.Lock(ctx, jwksLockID, time.Minute)
+		if err != nil {
+			logger.Warnw("unable to obtain lock", "lock", jwksLockID, "error", err)
```
Comment (sethvargo) [inline internal/jwks/handle_update_all.go:36]: Same about moving this after AlreadyLocked

File: internal/jwks/handle_update_all.go [hunk 2]
```diff
@@ -15,24 +15,45 @@
...
+			}
+		}()
+
 		if err := s.manager.UpdateAll(ctx); err != nil {
 			logger.Errorw("failed to update all", "error", err)
 			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 			return
 		}
 
+		mJWKSSuccess.M(1)
```
Comment (sethvargo) [inline internal/jwks/handle_update_all.go:56]: I think you need to call `stats.Record`?

File: internal/jwks/metrics.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		"Number of successful completions of jwks handler", stats.UnitDimensionless)
+)
+
+func init() {
+	observability.CollectViews([]*view.View{
+		{
+			Name:        metrics.MetricRoot + "jwks_success",
+			Description: "Number of successful completions of jwks handler",
+			Measure:     mJWKSSuccess,
+			Aggregation: view.Sum(),
```
Comment (sethvargo) [inline internal/jwks/metrics.go:37]: I think it's view.Count

File: internal/keyrotation/rotate.go [hunk 4]
```diff
@@ -67,6 +65,7 @@ func (s *Server) handleRotateKeys() http.Handler {
 			return
 		}
 
+		mRotationSuccess.M(1)
```
Comment (sethvargo) [inline internal/keyrotation/rotate.go:68]: stats.Record

File: internal/keyrotation/metrics.go [hunk 5]
```diff
@@ -45,5 +46,11 @@ func init() {
 			Measure:     mRevisionKeysDeleted,
 			Aggregation: view.Sum(),
 		},
+		{
+			Name:        metrics.MetricRoot + "rotation_success",
+			Description: "Number of successful completions of rotate handler",
+			Measure:     mRotationSuccess,
+			Aggregation: view.Sum(),
```
Comment (sethvargo) [inline internal/keyrotation/metrics.go:53]: I think you want count? 
 
> Count only counts the number of measurement points recorded. Distribution provides statistical summary of the aggregated data by counting how many recorded measurements fall into each bucket. Sum adds up the measurement values. LastValue just keeps track of the most recently recorded measurement value. All aggregations are cumulative.

File: internal/jwks/handle_update_all.go [hunk 6]
```diff
@@ -15,24 +15,46 @@
...
+			}
+		}()
+
 		if err := s.manager.UpdateAll(ctx); err != nil {
 			logger.Errorw("failed to update all", "error", err)
 			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 			return
 		}
 
+		stats.Record(ctx, mJWKSSuccess.M(1))
```
Comment (sethvargo) [inline internal/jwks/handle_update_all.go:57]: Handle this error?

File: internal/keyrotation/rotate.go [hunk 7]
```diff
@@ -67,6 +65,7 @@ func (s *Server) handleRotateKeys() http.Handler {
 			return
 		}
 
+		stats.Record(ctx, mRotationSuccess.M(1))
```
Comment (sethvargo) [inline internal/keyrotation/rotate.go:68]: Handle this error?

**Review by sethvargo — APPROVED** [review]
/lgtm

Two Qs about error handling that can be a follow-up

**Pull request merged**